### PR TITLE
UGRID-43: Handle SWIGTYPE_P_<T>

### PR DESCRIPTION
**Pull request by ahmad-el-sayed**

### CODE REVIEW

File: libs/UGridNET/dll/src/UGridReader.cs [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+            }
+        }
+
+        ~UGridReader()
+        {
+            // Cleanup in case Dispose wasn't called
+            Dispose(false);
+        }
+
+        public new void Dispose()
```
Comment (lucacarniato) [inline libs/UGridNET/dll/src/UGridReader.cs:41]: the base class provides the same implementation. is this implementation required?

File: libs/UGridNET/dll/src/UGridReader.cs [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+
+                base.Dispose(disposing);
+            }
+        }
+
+
+        private void ReadMesh1D()
+        {
+            int count = 0;
+            Invoke(() => UGrid.ug_topology_get_count(fileID, TopologyType.Mesh1dTopology, ref count));
```
Comment (lucacarniato) [inline libs/UGridNET/dll/src/UGridReader.cs:67]: consider a different name for Invoke. it is also a name of a commonly used c# function

File: libs/UGridNET/dll/src/UGridReader.cs [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+            {
+                for (int i = 0; i < count; i++)
+                {
+                    Mesh2D mesh2D = new Mesh2D();
+                    Invoke(() => UGrid.ug_mesh2d_inq(fileID, i, mesh2D));
+                    mesh2D.Allocate();
+                    Invoke(() => UGrid.ug_mesh2d_get(fileID, i, mesh2D));
+                    mesh2DList.Add(mesh2D);
+
+                    //var name = mesh2D.name.CopyToArray<byte>(UGrid.name_long_length);
```
Comment (lucacarniato) [inline libs/UGridNET/dll/src/UGridReader.cs:95]: in the final pr, consider removing commented code

File: libs/UGridNET/dll/src/UGridWrapper.cs [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+
+            int totalDimension = 1;
+            for (int i = 0; i < dimensionsCount; i++)
+            {
+                totalDimension *= dimensionVec[i];
+            }
+
+            T[] data = new T[totalDimension];
+
+            if (typeof(T) == typeof(double))
```
Comment (lucacarniato) [inline libs/UGridNET/dll/src/UGridWrapper.cs:137]: consider using a switch statement

**Review by lucacarniato — CHANGES_REQUESTED** [review]
looks good, fews suggestions for improvements and one possible redundancy in implementing the dispose method

File: libs/UGridAPI/src/UGrid.cpp [hunk 5]
```diff
@@ -51,22 +54,39 @@
...
         catch (const std::exception& e)
         {
-            strcpy(exceptionMessage, e.what());
+            std::strncpy(exceptionMessage, e.what(), max_chars_to_copy);
             return Exception;
         }
     }
 
-    static std::unique_ptr<ugrid::UGridEntity> get_topology(int file_id, int topology_id, int topology_type)
+    static std::unique_ptr<ugrid::UGridEntity> get_topology(int file_id,
```
[the remark this answers is not in the document]
Comment (ahmad-el-sayed) [reply libs/UGridAPI/src/UGrid.cpp:88]: Should be done in a refactoring PR.

File: libs/UGridAPI/src/UGrid.cpp [hunk 6]
```diff
@@ -92,7 +112,8 @@ namespace ugridapi
         auto const attributes = nc_var.getAtts();
         for (auto const& [name, value] : attributes)
         {
-            if (value.getType() == netCDF::NcType::nc_INT)
+            netCDF::NcType const type = value.getType();
+            if (type == netCDF::NcType::nc_INT)
```
[the remark this answers is not in the document]
Comment (ahmad-el-sayed) [reply libs/UGridAPI/src/UGrid.cpp:130]: Should be done in a refactoring PR.

File: libs/UGridNET/dll/src/UGridReader.cs (hunk 3 above)
Comment (ahmad-el-sayed) [reply libs/UGridNET/dll/src/UGridReader.cs]: It is a private method of a class in a namespace. It is fine.

File: libs/UGridNET/dll/src/UGridWrapper.cs (hunk 4 above)
Comment (ahmad-el-sayed) [reply libs/UGridNET/dll/src/UGridWrapper.cs]: done

File: libs/UGridNET/dll/src/Mesh1DExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+                {
+                    mesh1D.name = IntPtrHelpers.Allocate<byte>(UGrid.name_long_length);
+                    mesh1D.node_long_name = IntPtrHelpers.Allocate<byte>(UGrid.name_long_length * mesh1D.num_nodes);
+                    mesh1D.network_name = IntPtrHelpers.Allocate<byte>(UGrid.name_long_length);
+                    mesh1D.node_x = IntPtrHelpers.Allocate<double>(mesh1D.num_nodes);
+                    mesh1D.node_y = IntPtrHelpers.Allocate<double>(mesh1D.num_nodes);
+                    mesh1D.edge_x = IntPtrHelpers.Allocate<double>(mesh1D.num_edges);
+                    mesh1D.edge_y = IntPtrHelpers.Allocate<double>(mesh1D.num_edges);
+                    mesh1D.edge_nodes = IntPtrHelpers.Allocate<int>(mesh1D.num_edges * 2);
+                    mesh1D.edge_edge_id = IntPtrHelpers.Allocate<int>(mesh1D.num_nodes); // size?
```
[the remark this answers is not in the document]
Comment (ahmad-el-sayed) [reply libs/UGridNET/dll/src/Mesh1DExtensions.cs:26]: done

File: libs/UGridNET/dll/src/StringExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+            {
+                if (str == null)
+                {
+                    throw new ArgumentNullException(nameof(str), "String cannot be null.");
+                }
+
+                return Encoding.UTF8.GetBytes(str);
+            }
+
+            public static string PadRightUpTo(this string str, int maxLength)
```
[the remark this answers is not in the document]
Comment (ahmad-el-sayed) [reply libs/UGridNET/dll/src/StringExtensions.cs]: Turns out string class has PadRight. Extension was removed.

File: libs/UGridNET/dll/src/ContactsExtensions.cs [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+            /// <summary> Allocates unmanaged memory for all required properties of the <see cref="Contacts"/> instance. </summary>
+            /// <param name="contacts"> Instance of <see cref="Contacts"/>. </param>
+            /// <remarks> Memory allocated by this method must be freed by <see cref="Free"> </remarks>
+            public static void Allocate(this Contacts contacts)
+            {
+                try
+                {
+                    contacts.name = IntPtrHelpers.Allocate<byte>(UGrid.name_long_length);
+                    contacts.contact_name_id = IntPtrHelpers.Allocate<byte>(UGrid.name_length * contacts.num_contacts);
+                    contacts.mesh_from_name = IntPtrHelpers.Allocate<byte>(UGrid.name_long_length);
```
Comment (lucacarniato) [inline libs/UGridNET/dll/src/ContactsExtensions.cs:21]: UGrid standards use long_name. consider changing to long_name_length

File: libs/UGridNET/dll/src/IntPtrExtensions.cs [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+                }
+
+                if (array.Length == 0)
+                {
+                    throw new ArgumentException("Array is empty.", nameof(array));
+                }
+
+                switch (array)
+                {
+                    case int[] intArray:
```
Comment (lucacarniato) [inline libs/UGridNET/dll/src/IntPtrExtensions.cs:83]: if (typeof(T) == typeof(int))
    {
        Marshal.Copy((int[])(object)array, 0, pointer, array.Length);
    } this postpones the typecasting from the switch to Marshal.Copy. it is just another valid approach

**Review by lucacarniato — APPROVED** [review]

File: libs/UGridNET/dll/src/Mesh1DExtensions.cs [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+                    mesh1D.node_edge_id = IntPtrHelpers.Allocate<int>(mesh1D.num_nodes); // size?
```
[the remark this answers is not in the document]
Comment (ahmad-el-sayed) [reply libs/UGridNET/dll/src/Mesh1DExtensions.cs]: done

File: libs/UGridNET/dll/src/StringExtensions.cs [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+
+                return tokens;
+            }
+
+            public static List<List<string>> SplitString(string str, int tokenLength)
+            {
+                List<List<string>> tokens = new List<List<string>>();
+                int currentIndex = 0;
+
+                while (currentIndex < str.Length)
```
[the remark this answers is not in the document]
Comment (ahmad-el-sayed) [reply libs/UGridNET/dll/src/StringExtensions.cs:102]: This is not used. It was removed in a later commit.

File: libs/UGridNET/dll/src/ContactsExtensions.cs (hunk 9 above)
Comment (ahmad-el-sayed) [reply libs/UGridNET/dll/src/ContactsExtensions.cs:21]: OK. This was allocated following dimensions used in the API unit tests.

File: libs/UGridNET/dll/src/StringExtensions.cs [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+
+                return str.Substring(0, endIndex + 1);
+            }
+
+            public static List<string> SplitIntoSizedTokens(this string str, int tokenLength)
+            {
+                List<string> tokens = new List<string>();
+                int currentIndex = 0;
+
+                while (currentIndex < str.Length)
```
[the remark this answers is not in the document]
Comment (ahmad-el-sayed) [reply libs/UGridNET/dll/src/StringExtensions.cs]: done

File: libs/UGridNET/dll/src/UGridReader.cs (hunk 1 above)
Comment (ahmad-el-sayed) [reply libs/UGridNET/dll/src/UGridReader.cs]: reowrked

File: libs/UGridNET/dll/src/UGridBase.cs [hunk 14]
```diff
@@ -0,0 +1,180 @@
...
+        public ReadOnlyCollection<Mesh1D> Mesh1DList() { return mesh1DList.AsReadOnly(); }
+
+        [ExcludeFromCodeCoverage]
+        public ReadOnlyCollection<Mesh2D> Mesh2DList() { return mesh2DList.AsReadOnly(); }
+
+        [ExcludeFromCodeCoverage]
+        public ReadOnlyCollection<Contacts> ContactsList() { return contactsList.AsReadOnly(); }
+
+        [ExcludeFromCodeCoverage]
+        public ReadOnlyCollection<Network1D> Network1DList() { return network1DList.AsReadOnly(); }
```
Comment (ScottWillcox) [inline libs/UGridNET/dll/src/UGridBase.cs:109]: these could be properties

File: libs/UGridAPI/src/UGrid.cpp [hunk 15]
```diff
@@ -83,6 +103,19 @@ namespace ugridapi
         }
     }
 
+    /// @brief Clears and shrink strings consisting exclusively of mull characters
```
Comment (lucacarniato) [inline libs/UGridAPI/src/UGrid.cpp:106]: null

Comment (lucacarniato) [reply libs/UGridAPI/src/UGrid.cpp:106]: possible improvement 

static void ClearIfOnlyNullChars(std::string& str) 
{
    if (std::find_if(str.begin(), str.end(), [](char c) { return c != '\0'; }) != str.end())
        return;
    str.clear();
    str.shrink_to_fit();
}

File: libs/UGridAPI/src/UGrid.cpp [hunk 16]
```diff
@@ -92,22 +125,44 @@ namespace ugridapi
...
-            if (value.getType() == netCDF::NcType::nc_INT)
+            netCDF::NcType const type = value.getType();
+            if (type == netCDF::NcType::nc_INT)
             {
                 auto const attribute_length = value.getAttLength();
                 std::vector<int> items(attribute_length);
                 value.getValues(&items[0]);
                 std::stringstream os;
-                for (auto const& item : items)
+                for (auto it = items.begin(); it != items.end(); ++it)
```
Comment (lucacarniato) [inline libs/UGridAPI/src/UGrid.cpp:135]: if (!items.empty()) 
{
    std::copy(items.begin(), items.end() - 1, std::ostream_iterator<std::string>(os, " "));
    os << items.back();
}

File: libs/UGridNET/SWIG/interface/UGridTypemaps.i [hunk 17]
```diff
@@ -19,4 +24,251 @@
...
+
+// %apply int INPUT[] { int const* attribute_values } %{
+//     int ug_attribute_int_define(int file_id,
+//                                 char const* variable_name,
+//                                 char const* attribute_name,
+//                                 int const* attribute_values,
+//                                 int num_values);
+// %}
+
+// %apply double INPUT[] { double const* attribute_values } %{
```
Comment (lucacarniato) [inline libs/UGridNET/SWIG/interface/UGridTypemaps.i:58]: if this is commented code, remove it at the end

File: libs/UGridNET/test/src/ByteArrayExtensionsTests.cs [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+        [Test]
+        public void GetStringFromNullTerminatedArrayThrowsDueToEmptyBytesArray()
+        {
+            var bytes = new byte[] { };
+            string str;
+            Assert.Throws<ArgumentException>(() => str = bytes.GetStringFromNullTerminatedArray());
+        }
+
+        [Test]
+        public void GetStringFromNullTerminatedArrayThrowsDueToBytesArrayNotBeingNullTerminated()
```
Comment (lucacarniato) [inline libs/UGridNET/test/src/ByteArrayExtensionsTests.cs:29]: i suggest a more readable string: GetStringFromNullTerminatedArray_ShouldThrowArgumentException_WhenArrayIsNotNullTerminated

File: libs/UGridNET/test/src/UGridReaderTests.cs [hunk 19]
```diff
@@ -0,0 +1,287 @@
...
+        [Test]
+        public void GetEPSGCodeFromAVarWithValuelessAttribute()
+        {
+            var filePath = Path.Combine(TestDataPath, "OneMesh2D.nc");
+            UGridReader file = null;
+
+            try
+            {
+                file = new UGridReader(filePath);
+                var EPSGCode = file.GetEPSGCode();
```
Comment (lucacarniato) [inline libs/UGridNET/test/src/UGridReaderTests.cs:147]: EPSGCode int will cover most of the cases, the string version (e.g. EPSG_code = "EPSG:0") is an edge case we might not encounter

**Review by lucacarniato — APPROVED** [review]

File: libs/UGridNET/dll/src/ByteArrayExtensions.cs [hunk 20]
```diff
@@ -0,0 +1,46 @@
...
+            public static string GetStringFromNullTerminatedArray(this byte[] bytes, bool doTrimEnd = false)
+            {
+                if (bytes == null)
+                {
+                    throw new ArgumentNullException(nameof(bytes), "Byte array cannot be null.");
+                }
+
+                if (bytes.Length == 0)
+                {
+                    throw new ArgumentException("Byte array cannot be null.", nameof(bytes));
```
[the remark this answers is not in the document]
Comment (ahmad-el-sayed) [reply libs/UGridNET/dll/src/ByteArrayExtensions.cs:29]: copy pasta. gefixed.

File: libs/UGridNET/dll/src/ContactsExtensions.cs [hunk 21]
```diff
@@ -0,0 +1,53 @@
+using System;
+using System.Runtime.CompilerServices;
+
+[assembly: InternalsVisibleToAttribute("UGridNET.Tests")]
```
[the remark this answers is not in the document]
Comment (ahmad-el-sayed) [reply libs/UGridNET/dll/src/ContactsExtensions.cs]: You are right but Directory.Build.props is a CMake-configurable file that used by 2 solutions. This should be changed in the future.

File: libs/UGridNET/dll/src/IntPtrExtensions.cs [hunk 22]
```diff
@@ -0,0 +1,98 @@
...
+            /// </summary>
+            /// <typeparam name="T">The type of elements in the array. Must be a value type.</typeparam>
+            /// <param name="pointer">The unmanaged memory pointer.</param>
+            /// <param name="length">The number of elements to copy.</param>
+            /// <returns>A managed array containing the copied data.</returns>
+            /// <exception cref="ArgumentNullException">Thrown when the pointer is null.</exception>
+            /// <exception cref="ArgumentException">Thrown when the length is less than or equal to zero.</exception>
+            /// <exception cref="NotSupportedException">Thrown when the type of elements is not supported.</exception>
+            public static T[] CopyToArray<T>(this IntPtr pointer, int length) where T : struct
+            {
```
[the remark this answers is not in the document]
Comment (ahmad-el-sayed) [reply libs/UGridNET/dll/src/IntPtrExtensions.cs:24]: This is handled by the default case but maybe it is better do it before `var array = new T[length];`

File: libs/UGridAPI/include/UGridAPI/UGrid.hpp [hunk 23]
```diff
@@ -65,32 +73,10 @@ namespace ugridapi
             Exception = 1,
```
[the remark this answers is not in the document]
Comment (ahmad-el-sayed) [reply libs/UGridAPI/include/UGridAPI/UGrid.hpp:73]: That's inherited. This is not a type, simply an enum indicating failure upon handling an exception.

File: libs/UGridNET/dll/src/UGridBase.cs (hunk 14 above)
Comment (ahmad-el-sayed) [reply libs/UGridNET/dll/src/UGridBase.cs:109]: Done

File: libs/UGridNET/dll/src/AssemblyInfo.cs
[the remark this answers is not in the document]
Comment (ahmad-el-sayed) [reply libs/UGridNET/dll/src/AssemblyInfo.cs:1]: The csproj is not part of the repo, it is rather generated by CMake in the build dir. So editing the csproj to add the above is not possble. _Thocccc..._

Target properties are set as such 
```
set_target_properties(
  ${TARGET_NAME}
  PROPERTIES
    DOTNET_TARGET_FRAMEWORK "netstandard2.0"
    DOTNET_SDK "Microsoft.NET.Sdk"
    VS_GLOBAL_PlatformTarget "AnyCPU"
)
```

Unfortunately,  I could not find a property for setting internal visibility.

File: cmake/find_packages.cmake [hunk 24]
```diff
@@ -1,44 +1,93 @@
...
-endif()
-
-# git
-find_package(Git)
-if (Git_FOUND)
-    message(STATUS "Found Git ${GIT_VERSION_STRING}")
-endif()
-
+# Boost
+set(BOOST_MIN_VERSION "1.78.0")
```
[the remark this answers is not in the document]
Comment (ahmad-el-sayed) [reply cmake/find_packages.cmake:2]: I agree. The readme needs a lot of improvements.  When it comes to dependencies, It lists boost as the only dependency.  There are further optional CMake config params that can be used. I suggest doing this in a dedicated issue. Filed here [UGRID-44](https://issuetracker.deltares.nl/browse/UGRID-44).

File: cmake/find_packages.cmake [hunk 25]
```diff
@@ -1,44 +1,93 @@
...
+    endif()
+    get_target_property(IMPORTED_TARGET_LOCATION ${IMPORTED_TARGET_NAME} LOCATION)
+    message(STATUS "Runtime dependency ${IMPORTED_TARGET_NAME} (${REQUIRED_IMPORTED_TARGET_TYPE}): ${IMPORTED_TARGET_LOCATION}")
+endfunction()
+
+check_runtime_dependency(netCDF::netcdf SHARED_LIBRARY)
+check_runtime_dependency(netCDF::netcdf-cxx4 SHARED_LIBRARY)
+check_runtime_dependency(hdf5-shared SHARED_LIBRARY)
+check_runtime_dependency(hdf5_hl-shared SHARED_LIBRARY)
+check_runtime_dependency(ZLIB::ZLIB UNKNOWN_LIBRARY) #SHARED_LIBRARY)
```
[the remark this answers is not in the document]
Comment (ahmad-el-sayed) [reply cmake/find_packages.cmake:92]: done

File: libs/UGridAPI/src/UGrid.cpp (hunk 15 above)
Comment (ahmad-el-sayed) [reply libs/UGridAPI/src/UGrid.cpp:106]: The function checks if the string consists of only null characters.

File: libs/UGridNET/SWIG/interface/UGridTypemaps.i (hunk 17 above)
Comment (ahmad-el-sayed) [reply libs/UGridNET/SWIG/interface/UGridTypemaps.i:58]: done

File: libs/UGridNET/dll/src/ByteArrayExtensions.cs [hunk 26]
```diff
@@ -0,0 +1,43 @@
...
+                }
+
+                if (bytes[bytes.Length - 1] != 0x00)
+                {
+                    throw new ArgumentException("Byte array is not null-terminated.", nameof(bytes));
+                }
+
+                string str = Encoding.UTF8.GetString(bytes, 0, bytes.Length - 1);
```
[the remark this answers is not in the document]
Comment (ahmad-el-sayed) [reply libs/UGridNET/dll/src/ByteArrayExtensions.cs:37]: Yes, UTF-8 is sufficient in this project as far as `char*` on the C side is concerned. Beside, it is consistent across all platforms.

File: libs/UGridAPI/src/UGrid.cpp (hunk 16 above)
Comment (ahmad-el-sayed) [reply libs/UGridAPI/src/UGrid.cpp:135]: Should be done in a refactoring issue. Here an existing issue was fixed with minimal changes. Newly added code follows old logic.

File: libs/UGridNET/test/src/UGridReaderTests.cs (hunk 19 above)
Comment (ahmad-el-sayed) [reply libs/UGridNET/test/src/UGridReaderTests.cs:147]: Code 0 is not even valid. However,  the file uses in the test has it. The generating tool is not checking the range of valid epsg codes and likely defaulting to the default value of the integer type.

This a test. If the expected value is retrieved then the test serves its own purpose.
